xonsh/ptk_shell/shell.py

```diff
@@ -298,58 +355,12 @@ def singleline(
             "complete_in_thread": complete_in_thread,
         }
 
-        if env.get("COLOR_INPUT"):
-            events.on_timingprobe.fire(name="on_pre_prompt_style")
-            style_overrides_env = env.get("PTK_STYLE_OVERRIDES", {}).copy()
-            if (
-                len(style_overrides_env) > 0
-                and not self._overrides_deprecation_warning_shown
-            ):
-                print_warning(
-                    "$PTK_STYLE_OVERRIDES is deprecated, use $XONSH_STYLE_OVERRIDES instead!"
-                )
-                self._overrides_deprecation_warning_shown = True
-            style_overrides_env.update(env.get("XONSH_STYLE_OVERRIDES", {}))
-
-            if HAS_PYGMENTS:
-                prompt_args["lexer"] = PygmentsLexer(pyghooks.XonshLexer)
-                style = _style_from_pygments_cls(
-                    pyghooks.xonsh_style_proxy(self.styler)
-                )
-                if len(self.styler.non_pygments_rules) > 0:
-                    try:
-                        style = merge_styles(
-                            [
-                                style,
-                                _style_from_pygments_dict(
-                                    self.styler.non_pygments_rules
-                                ),
-                            ]
-                        )
-                    except (AttributeError, TypeError, ValueError) as style_exception:
-                        print_warning(
-                            f"Error applying style override!\n{style_exception}\n"
-                        )
-
-            else:
-                style = _style_from_pygments_dict(DEFAULT_STYLE_DICT)
-
-            if len(style_overrides_env) > 0:
-                try:
-                    style = merge_styles(
-                        [style, _style_from_pygments_dict(style_overrides_env)]
-                    )
-                except (AttributeError, TypeError, ValueError) as style_exception:
-                    print_warning(
-                        f"Error applying style override!\n{style_exception}\n"
-                    )
-
-            prompt_args["style"] = style
-            events.on_timingprobe.fire(name="on_post_prompt_style")
-
         if env["ENABLE_ASYNC_PROMPT"]:
             # once the prompt is done, update it in background as each future is completed
             prompt_args["pre_run"] = self.prompt_formatter.start_update
```

Comment:
Seems to me calling `add_attr` from the `pre_fun` will run it lazily. Can you try that to see if it works?

Doing it in another thread isn't really a good solution because it isn't io-bound and will do more harm than good (https://stackoverflow.com/questions/62154869/multi-threading-and-single-threading-performance-issues-in-cpu-bound-task)
Also running in it a thread isn't really _lazy_, but _parallel_ (and won't really be parallel because of the GIL)

Reply:
it is like running it later. Since it is an interactive prompt, it is going to wait for user input at later point. This is just choosing to run certain things later. I know it is not parallel. 

> Seems to me calling add_attr from the pre_fun will run it lazily. Can you try that to see if it works?

add_attr ?

Reply:
Instead of submitting `self.add_attrs` to the threadpool in `start`, set `ptk_kwargs['pre_run']` to a function which calls `self.add_attrs` and `self.prompt_formatter.start_update`.
I think this will call it lazily as needed (not sure though)

Reply:
`pre_run` is called in the same thread. The idea here is to offload the style related call to another thread. Because it is not essential for the prompt to start. Once the user starts typing we will be seeing the color. 

Also this asynchronous prompt sets style per session. Otherwise it is set per prompt.